sdks/java/core/src/main/java/org/apache/beam/sdk/util/ReduceFnRunner.java

```diff
@@ -856,21 +875,33 @@ private Instant scheduleEndOfWindowOrGarbageCollectionTimer(
         directContext.window(),
         inputWM,
         timerInternals.currentOutputWatermarkTime());
+    Preconditions.checkState(!fireTime.isAfter(BoundedWindow.TIMESTAMP_MAX_VALUE),
+                             "Fire time %s is beyond end-of-time", fireTime);
     directContext.timers().setTimer(fireTime, TimeDomain.EVENT_TIME);
     return fireTime;
   }
 
-  private void cancelEndOfWindowAndGarbageCollectionTimers(ReduceFn<?, ?, ?, W>.Context context) {
+  private void cancelEndOfWindowAndGarbageCollectionTimers(
+      ReduceFn<?, ?, ?, W>.Context directContext) {
     WindowTracing.debug(
         "ReduceFnRunner.cancelEndOfWindowAndGarbageCollectionTimers: Deleting timers for "
         + "key:{}; window:{} where inputWatermark:{}; outputWatermark:{}",
-        key, context.window(), timerInternals.currentInputWatermarkTime(),
+        key, directContext.window(), timerInternals.currentInputWatermarkTime(),
         timerInternals.currentOutputWatermarkTime());
-    Instant timer = context.window().maxTimestamp();
-    context.timers().deleteTimer(timer, TimeDomain.EVENT_TIME);
-    if (windowingStrategy.getAllowedLateness().isLongerThan(Duration.ZERO)) {
-      timer = timer.plus(windowingStrategy.getAllowedLateness());
-      context.timers().deleteTimer(timer, TimeDomain.EVENT_TIME);
+    Instant eow = directContext.window().maxTimestamp();
+    directContext.timers().deleteTimer(eow, TimeDomain.EVENT_TIME);
+    Instant gc = windowMaxTimestampPlusAllowedLateness(directContext.window());
+    if (gc.isAfter(eow)) {
+      directContext.timers().deleteTimer(eow, TimeDomain.EVENT_TIME);
+    }
+  }
+
+  private Instant windowMaxTimestampPlusAllowedLateness(W window) {
```

Comment:
Strange to abstract this logic out and then give a name that allows no abstraction. And since it does not allow abstraction, the definition is incorrect relative to the name.

Perhaps just `expirationTimestamp` or some such.